### PR TITLE
Enable parsing columns from file path for Broker Load (#1582)

### PR DESCRIPTION
**Pull request by yuanlihan**
Currently, we do not support parsing encoded/compressed columns in file path, eg: extract column k1 from file path `/path/to/dir/k1=1/xxx.csv`

This patch is able to parse columns from file path like in Spark(Partition Discovery).

This patch parse partition columns at BrokerScanNode.java and save parsing result of each file path as a property of TBrokerRangeDesc, then the broker reader of BE can read the value of specified partition column.

### CODE REVIEW

File: be/src/exec/broker_scanner.cpp [hunk 1]
```diff
@@ -489,18 +504,18 @@ bool BrokerScanner::line_to_src_tuple(const Slice& line) {
...
 
-    for (int i = 0; i < values.size(); ++i) {
+    int file_column_index = 0;
+    for (int i = 0; i < _src_slot_descs.size(); ++i) {
         auto slot_desc = _src_slot_descs[i];
-        const Slice& value = values[i];
-        if (slot_desc->is_nullable() && is_null(value)) {
-            _src_tuple->set_null(slot_desc->null_indicator_offset());
-            continue;
+        auto iter = _columns_from_path.find(slot_desc->col_name());
```
Comment (imay) [inline be/src/exec/broker_scanner.cpp]: Could you find this one time when this class is initialized?
If you did like that, we can reduce find operation here for each filed of each row.

File: be/src/exec/broker_scanner.cpp [hunk 2]
```diff
@@ -489,18 +504,18 @@ bool BrokerScanner::line_to_src_tuple(const Slice& line) {
...
+        if (iter != _columns_from_path.end()) {
+            std::string partitioned_field = iter->second;
+            const Slice value = Slice(partitioned_field.c_str(), partitioned_field.size());
```
Comment (imay) [inline be/src/exec/broker_scanner.cpp]: What you reference `partitioned_field` is a temporary variable. It is not valid out of code block.

File: be/src/exec/parquet_reader.cpp [hunk 3]
```diff
@@ -204,13 +208,23 @@ Status ParquetReaderWrap::read(Tuple* tuple, const std::vector<SlotDescriptor*>&
...
     const uint8_t *value = nullptr;
+    int index = 0;
     int column_index = 0;
     try {
         size_t slots = tuple_slot_descs.size();
         for (size_t i = 0; i < slots; ++i) {
             auto slot_desc = tuple_slot_descs[i];
-            column_index = i;// column index in batch record
-            switch (_parquet_column_type[i]) {
+            auto iter = _columns_from_path.find(slot_desc->col_name());
```
Comment (imay) [inline be/src/exec/parquet_reader.cpp]: I think we can finish it in BrokerScanner's line_to_src_tuple. Which can avoid same code in different places.

File: fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java [hunk 4]
```diff
@@ -696,9 +699,40 @@ private TBrokerRangeDesc createBrokerRangeDesc(long curFileOffset, TBrokerFileSt
...
 
+    private Map<String, String> parseColumnsFromPath(String filePath, List<String> columnsFromPath) throws UserException {
+        if (columnsFromPath == null || columnsFromPath.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        String[] strings = filePath.split("/");
+        Map<String, String> columns = new HashMap<>();
+        for (int i = strings.length - 1; i >= 0; i--) {
+            String str = strings[i];
+            if (str == null || str.isEmpty() || !str.contains("=")) {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java]: When we meet a path like "/xx/yy/zz/k1=1/aa",  I think we should think this path is not a valid path. We only accept paths like "xx/yy/zz/aa/k1=1"

File: fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java [hunk 5]
```diff
@@ -696,9 +699,40 @@ private TBrokerRangeDesc createBrokerRangeDesc(long curFileOffset, TBrokerFileSt
...
+            }
+            columns.put(pair[0], pair[1]);
+            if (columns.size() > columnsFromPath.size()) {
+                break;
+            }
+        }
+        if (columns.size() != columnsFromPath.size()) {
+            throw new UserException("Fail to parse columnsFromPath, expected: " + columnsFromPath + ", filePath: " + filePath);
+        }
+        LOG.info("columns from path: " + columns + ", filePath: " + filePath);
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java]: You should remove this log

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/exec/parquet_reader.cpp (hunk 3 above)
Comment (yuanlihan) [reply be/src/exec/parquet_reader.cpp]: Currently, we have two broker scanners. One is BrokerScanner using for reading CSV like row-storage sources. The other is ParquetScanner using for read parquet files, which is a column-storage source. It seems that we cannot reuse the `line_to_src_tuple` method of BrokerScanner

File: fe/src/main/cup/sql_parser.cup [hunk 6]
```diff
@@ -1050,10 +1050,11 @@ data_desc ::=
     opt_partitions:partitionNames
     opt_field_term:colSep
     opt_file_format:fileFormat
+    opt_columns_from_path:columnsFromPath
```
[the remark this answers is not in the document]
Comment (yuanlihan) [reply fe/src/main/cup/sql_parser.cup]: > Could you please put this behind colList?

Sure

File: fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java (hunk 4 above)
Comment (yuanlihan) [reply fe/src/main/java/org/apache/doris/planner/BrokerScanNode.java]: > When we meet a path like "/xx/yy/zz/k1=1/aa", I think we should think this path is not a valid path. We only accept paths like "xx/yy/zz/aa/k1=1"

I will raise an exception here.

File: be/src/exec/parquet_reader.cpp (hunk 3 above)
Comment (imay) [reply be/src/exec/parquet_reader.cpp]: I think we can write a common function to fill partition columns which can be used in both two scanners, like `fill_dest_tuple`. If this function is not easy to write and used by scanners, it's OK to write one copy for each.

**Pull request closed without merging**